Add blog post on MongoDB sorting for scalar values (#3200)
FerretDB/FerretDB: website/blog/2023-10-25-mongodb-sorting-scalar.md

---
slug: mongodb-sorting-scalar
title: How MongoDB Sorting Works for Scalar Values
authors: [chi]
description: >
  In this blog post, we explore how MongoDB sorting works for scalar values.
keywords:
  - mongodb sorting
  - bson comparison order
image: /img/blog/post-cover-image.jpg

> tags: [community, product, tutorial]

Author: @Fashander are these tags sensible? I kept keywords let me know if you think they are note necessary.

Author: Removed keywords.

---
slug: mongodb-sorting-scalar
title: How MongoDB Sorting Works for Scalar Values
authors: [chi]
description: >
  In this blog post, we explore how MongoDB sorting works for scalar values.
image: /img/blog/mongodb-sorting-scalar.jpg
tags: [community, product, tutorial]
---

![How MongoDB Sorting Works for Scalar Values](/img/blog/mongodb-sorting-scalar.jpg)

In this blog post, we delve into the process of sorting scalar values in MongoDB.

<!--truncate-->

Sorting in MongoDB involves comparing BSON values to ascertain their relative order – whether one value is equal to, greater than, or less than another.

> The resultant sorted array can be in either ascending or descending order.

Reviewer: Yes. I focused mainly on the content so it flows better so If you see any place that requires technical changes please update them.

  { _id: 2, name: 'sandals', size: 9, color: null },
  { _id: 4, name: 'sneakers', size: 8.5, color: 'blue' },
  { _id: 3, name: 'boots', size: 8, color: 'black' },
  { _id: 5, name: 'slippers' }
]
```

This time, the output is sorted first by `flip flops` with String `size` field, then by `size` field with Numbers `sandals`, `sneakers` and `boots` and finally `slippers` with a non-existent `size` field.

### Using `_id` as the second sort field

Suppose you want to sort the documents by the `color` field.
You encounter multiple documents with the color `blue`, and one document has a Null value for this field while another is missing it altogether.

For instance, `flip flops` has a Null value in the `color` field, and the `slippers` document lacks this field.
Since Null and non-existent fields are considered equivalent in sorting, either could appear first.
In situations like this, the default order in which the records were retrieved from the database is applied.

To maintain a consistent sort order, it's advised to use `_id` as a secondary sorting option.
In this setup, if multiple documents have the same or equivalent `color` values, it will rely on the `_id` field for sorting.

> The unique nature of the `_id` field ensures that the sorted output remains consistent.

Author: ```suggestion
The `_id` field value is unique within the collection which ensures that the sorted output remains consistent.
```

@Fashander trying to emphasis that `_id` is unique.
